application/Profile/DBProfile.py

```diff
@@ -118,42 +143,56 @@ class Session(db.Model):
 
             @validates('nickname')
             def validate_nickname(self, key, nickname):
-                assert len(nickname) <= 8, \
-                    'Entered nickname is too long'
+                try:
+                    assert len(nickname) <= 8
+                except AssertionError as ae:
+                    logger.error('Entered nickname is too long')
+                    raise ae
+
                 return nickname
 
             username = db.Column(db.String, nullable=False)
             private_key = db.Column(db.Text, nullable=True)
 
+            logger.info("Defined a database table: Session")
+
         class VNCCredentials(db.Model):
             __table_args__ = {"schema": "ictrl"}
 
             session_id = db.Column(UUID(as_uuid=True), db.ForeignKey('ictrl.session.id'), primary_key=True,
                                    nullable=False)
             credentials = db.Column(db.Text, nullable=False)
 
+            logger.info("Defined a database table: VNCCredentials")
+
         self.User = User
         self.Session = Session
         self.VNCCredentials = VNCCredentials
 
         # db.drop_all()
         db.engine.execute("CREATE SCHEMA IF NOT EXISTS ictrl;")
         db.create_all()
+        logger.info("Created database SCHEMA ictrl and created all databases defined")
 
     def login(self, username, password):
         username = username.lower()
         password_bytes = password.encode('ascii')
 
         user = self.User.query.filter_by(username=username).first()
-        if user is None:
-            abort(403, 'ACCOUNT_WRONG_USERNAME')
+        hashed_password_bytes = user.password.encode('ascii')
+        if (user is None) or (not bcrypt.checkpw(password_bytes, hashed_password_bytes)):
+            #Reason being is for security concern, we want to be as ambiguous as possible
+            #In a scenario a hacker wants to hack into an account
+            #You wouldn't want him to know whether your username or your password is correct
+            #If he guesses the username then he can guess the password next, given by the original error messages
+            error_msg = 'ACCOUNT_WRONG_USERNAME or ACCOUNT_WRONG_PASSWORD'
+            logger.error(error_msg)
+            abort(403, error_msg)
```

Comment:
This is some good thinking but the accuracy in the error message was intended back then (when the web service was in beta phase and opened to uoft emails only, and occasionally we have student users who could not remember whether they have registered an account). 

Thinking the service should eventually scale and open to the public, I agree we should use a unified error code for both errors to prevent hackers from obtaining our user information. However, that also means we need to make the corresponding changes in the web frontend as well: https://github.dev/junhaoliao/iCtrl/blob/02b56d8ba0801577155f8bd3cfafbdf7397eef0d/client/src/interface/components/LogIn/index.js#L151

Also, the password check was intentionally delayed after verifying whether the user is activated (verified their own email) or not to avoid DDOS attacks, as hashing computation is unusually expensive.

Again I appreciate the thoughts put into this. For the unified error code / proper user authentication implementation, can we treat it as a separate issue and handle in anther PR? What are your thoughts on this?

Reply:
Sure, this is a separate issue so I can create another branch to make a change on this